### ISSUE TITLE
manhole telnet server broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2099]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2099|
|Type|defect|
|Created|2006-09-17 23:31:54Z|

Merge of [#1967](https://github.com/twisted/twisted/issues/1967) branch looks to have broken it for the linux telnet client.

<details><summary>Searchable metadata</summary>

```
trac-id__2099 2099
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__Conch_0_9 Conch-0.9
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__conch conch
keywords__ 
time__1158535914000000 1158535914000000
changetime__1190290619000000 1190290619000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Started on this in manhole-echo-negotiation-2099.

I may revert [#1967](https://github.com/twisted/twisted/issues/1967), fold it in with this branch, and then re-merge the whole thing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

This sounds like a regression.  Should it be blocking 2.5?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

It should block Conch 0.8.  If Conch 0.8 should block Twisted 2.5, there should be some other way of recording that fact.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [19153]) Revert r17829

Refs [#1967](https://github.com/twisted/twisted/issues/1967)
Refs [#2099](https://github.com/twisted/twisted/issues/2099)

This changeset introduced a regression in untested code which manifests
when accessing a conch telnet server using the NetKit telnet client.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

Uunnnnnnnnggggghhhhhhhhhhhh.  I don't know what the regression was and I can't reproduce it with the revision of trunk which had this branch merged.

For anyone watching, this isn't a very good ticket:

  * it's for a regression, the branch which introduced the regression should have been reverted and the ticket for it re-opened with an explanation of the problem, a new ticket should not have been created.
  * it doesn't actually explain what the regression _was_.

As penance I re-implemented the fix for [#1967](https://github.com/twisted/twisted/issues/1967) from scratch and tested the program attached there on both linux and windows, and tested ```twistd -n manhole --telnetPort 2323 --passwd <(echo username: password)``` on linux and windows.  The former works on both, the latter only works on linux, both in trunk and with the new [[#1967](https://github.com/twisted/twisted/issues/1967)](https://github.com/twisted/twisted/issues/1967) branch.  The problem on windows seems to be related to echo option negotiation.  The app does actually sort of work, but you can't see anything you type until you submit it to the server.  This is a separate issue, I think.

Closing this, since there won't be any code related to it, only the new branch for #1967.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [21299]) Merge teloptneg-1967

Author: exarkun
Reviewer: therve
Fixes [#1967](https://github.com/twisted/twisted/issues/1967)
Refs [#2099](https://github.com/twisted/twisted/issues/2099)

Change the base telnet protocol implementation, twisted.conch.telnet.Telnet,
so that, by default, it cleanly rejects all attempts to enable options, rather
than raising AttributeErrors which cause the connection to be lost.  Also
change it to raise NotImplementedError, rather than an AttributeError, to
indicate a programming error if one of its disable callbacks is somehow
invoked.

This allows simple Telnet subclasses which do not customize option negotiation
to work when used with clients which attempt option negotiation.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
